spec/Section 6 -- Execution.md

```diff
@@ -506,26 +561,40 @@ CollectFields(objectType, selectionSet, variableValues, visitedFragments):
       * If {DoesFragmentTypeApply(objectType, fragmentType)} is false, continue
         with the next {selection} in {selectionSet}.
       * Let {fragmentSelectionSet} be the top-level selection set of {fragment}.
+      * If {fragmentSpreadName} provides the directive `@defer`, let {deferDirective} be that directive.
+        * If {deferDirective}'s {if} argument is {true} or is a variable in {variableValues} with the value {true}:
+          * Let {deferredFragment} be the result of calling {DeferFragment(objectType, objectValue, fragmentSelectionSet, parentPath)}.
+          * Append {deferredFragment} to {deferredFragments}.
+          * Continue with the next {selection} in {selectionSet}.
       * Let {fragmentGroupedFieldSet} be the result of calling
         {CollectFields(objectType, fragmentSelectionSet, variableValues, visitedFragments)}.
       * For each {fragmentGroup} in {fragmentGroupedFieldSet}:
         * Let {responseKey} be the response key shared by all fields in {fragmentGroup}.
         * Let {groupForResponseKey} be the list in {groupedFields} for
           {responseKey}; if no such list exists, create it as an empty list.
         * Append all items in {fragmentGroup} to {groupForResponseKey}.
+      * Add {fragmentSpreadName} to {visitedFragments}.
```

Comment:
Doesn't this change open the possibility of infinite recursion because you visit the fragment before you mark that it was visited?

Reply:
i updated this a bit and now only fragments that have `@defer` will skip the check for visited fragments, which does leave open the possibility of infinite recursion, but I think that should be caught by validation rules? Another option is to make this into something that keys on both the fragment name and the defer directive.